lms/djangoapps/lti_provider/outcomes.py

```diff
@@ -0,0 +1,199 @@
+        assert consumer_key, "oauth_consumer_key is not an optional parameter."
+
+        # Both usage and course ID parameters are supplied in the LTI launch URL
+        usage_key = request_params.get('usage_key', None)
+        course_key = request_params.get('course_key', None)
+        assert None not in [usage_key, course_key], "usage_key ({}) and course_key ({}) should not be None".format(
+            usage_key, course_key
+        )
+
+        # Create a record of the outcome service if necessary
+        outcomes, __ = OutcomeService.objects.get_or_create(
+            lis_outcome_service_url=result_service,
+            consumer_key=consumer_key
+        )
+        # There could be a case where an earlier launch registered the outcome
+        # service but did not have the consumer instance guid available, in
+        # which case the outcome service would only have the consumer key
+        # available. If we now have better information, we can improve that
+        # record.
+        if instance_guid and not outcomes.instance_guid:
+            outcomes.instance_guid = instance_guid
+            outcomes.save()
+
+        # Create a record for this assignment. Note that there may already
+        # be a record, if this assignment has been launched more than once
+        # by the same user/tool consumer combination.
+        GradedAssignment.objects.get_or_create(
+            lis_result_sourcedid=result_id,
+            course_key=course_key,
+            usage_key=usage_key,
+            user=user,
+            outcome_service=outcomes
+        )
+
+
+# Pylint doesn't recognize members in the LXML module
+# pylint: disable=no-member
+def generate_replace_result_xml(result_sourcedid, score):
+    """
+    Create the XML document that contains the new score to be sent to the LTI
+    consumer. The format of this message is defined in the LTI 1.1 spec.
+    """
+    envelope = etree.Element(
+        'imsx_POXEnvelopeRequest',
+        xmlns='http://www.imsglobal.org/services/ltiv1p1/xsd/imsoms_v1p0'
+    )
+    header = etree.SubElement(envelope, 'imsx_POXHeader')
+    header_info = etree.SubElement(header, 'imsx_POXRequestHeaderInfo')
+    version = etree.SubElement(header_info, 'imsx_version')
+    message_id = etree.SubElement(header_info, 'imsx_messageIdentifier')
+    body = etree.SubElement(envelope, 'imsx_POXBody')
+    replace_result_request = etree.SubElement(body, 'replaceResultRequest')
+    result_record = etree.SubElement(replace_result_request, 'resultRecord')
+    sourced_guid = etree.SubElement(result_record, 'sourcedGUID')
+    sourced_id = etree.SubElement(sourced_guid, 'sourcedId')
+    result = etree.SubElement(result_record, 'result')
+    result_score = etree.SubElement(result, 'resultScore')
+    language = etree.SubElement(result_score, 'language')
+    text_string = etree.SubElement(result_score, 'textString')
+
+    version.text = 'V1.0'
+    message_id.text = str(uuid.uuid4())
```

Comment:
Does this just need to be unique and reasonably random, or are there crypto-level requirements about the randomness for this?